path/path.go

```diff
@@ -3,188 +3,301 @@ package path
 
 import (
 	"fmt"
-	"path"
+	gopath "path"
 	"strings"
 
-	cid "github.com/ipfs/go-cid"
+	"github.com/ipfs/go-cid"
+	"github.com/libp2p/go-libp2p/core/peer"
 )
 
-// A Path represents an ipfs content path:
-//   - <cid>/path/to/file
-//   - /ipfs/<cid>
-//   - /ipns/<cid>/path/to/folder
-//   - etc
-type Path string
-
-// ^^^
-// TODO: debate making this a private struct wrapped in a public interface
-// would allow us to control creation, and cache segments.
+const (
+	IPFSNamespace = "ipfs"
+	IPNSNamespace = "ipns"
+	IPLDNamespace = "ipld"
+)
 
-// FromString safely converts a string type to a Path type.
-func FromString(s string) Path {
-	return Path(s)
+// Path is a generic path. Paths must be prefixed with a valid prefix.
+type Path interface {
+	// String returns the path as a string.
+	String() string
+
+	// Namespace returns the first component of the path. For example, the namespace
+	// of "/ipfs/bafy" is "ipfs".
+	Namespace() string
+
+	// Mutable returns false if the data pointed to by this path is guaranteed to not
+	// change. Note that resolved mutable paths can be immutable.
+	Mutable() bool
+
+	// Root returns the [cid.Cid] of the root object of the path. Root can return
+	// [cid.Undef] for Mutable IPNS paths that use [DNSLink].
+	//
+	// [DNSLink]: https://dnslink.dev/
+	Root() cid.Cid
+
+	// Segments returns the different elements ofd a path, which are delimited
+	// by a forward slash ("/"). The leading slash must be ignored, that is, no
+	// segment should be empty.
+	Segments() []string
 }
 
-// FromCid safely converts a cid.Cid type to a Path type.
-func FromCid(c cid.Cid) Path {
-	return Path("/ipfs/" + c.String())
+// ResolvedPath is a [Path] which was resolved to the last resolvable node.
+type ResolvedPath interface {
```

Comment:
Maybe out of scope for this PR, but hopefully not. I'd like to better document what "last resolvable node" means.

For example:
1. A path that's been fully resolved down to its final IPLD block (i.e. the remainder when traversed does not cross any block boundaries)
2. A path that was resolved as far as the implementation could (e.g. it resolved DNSLink but failed at the IPNS resolution, it resolved a dag-cbor part of the path but couldn't understand a dag-jose part of the path, it couldn't find/access some block in the path, etc.)

It seems like this tries to imply the former, although if true I wonder how much this resolution is particularly needed in comparison to others like:
- I started with `/ipns/` and I want the immutable path
- I want the root CIDs of all the "entities" being traversed
- I want all of the ipns resolution information AND the root CIDs of the entities being traversed

Understanding where/why this is used might help us come up with something better here (and need fewer path wrappers or extra lookups in places like the gateway code).

Reply:
@aschmahmann Indeed, and I do also think that the current meaning is **2**. I checked and the only places across both Boxo and Kubo where it is used is:

- `gateway.ContentPathMetadata`, to represent the `LastSegment`, which I think we only do because it's the type used by the following two:
  - `kubo/client/rpc.ResolvePath` 
  - `kubo/coreapi.NewResolvedPath` 

I think we have the opportunity here to get rid of `ResolvedPath`. From what I can see, a path, in its generic form, is just `/{namespace}/{root}[/path]`, where:

- `namespace` can be `ipfs`, `ipns` or `ipld`.
- `root` can be a CID for all, can be a `ipns.Name` for `ipns` and a FQDN for `ipns` (DNSLink).

The key thing to note is that a `ResolvedPath` `Cid` is not the same as `Root`. `Cid` returns the CID of the last resolvable block (see `path/resolver.Resolver`).


cc @Jorropo because you also had questions about ResolvedPath vs. ImmutablePath

Reply:
See [0a71844](https://github.com/ipfs/boxo/pull/334/commits/0a718449a6f1a3a8f8646783e07aa028dbddecd0), where I removed `ResolvedPath` and use `ImmutablePath`. I think this seems like the way to go.